File: apple/Sources/ReactNativeWatchOS/Modifiers/ScrollDismissesKeyboard.swift
import SwiftUI

/// SwiftUI `.scrollDismissesKeyboard(_:)`. Configures how a scroll view
/// dismisses the keyboard during interaction.
///
/// Gated to watchOS 10+: `ScrollDismissesKeyboardMode` is only reliably
/// constructible on watchOS 10+, so on watchOS 9 the view is returned
/// unchanged (the system default dismiss behavior still applies).
enum RNWScrollDismissesKeyboardModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("scrollDismissesKeyboard") { view, params, _ in
            if #available(watchOS 10.0, *) {
                return AnyView(view.scrollDismissesKeyboard(mode(params.string("mode"))))
            }
            return view
        }
    }

    @available(watchOS 10.0, *)
    private static func mode(_ s: String?) -> ScrollDismissesKeyboardMode {
        switch s {
        case "immediately":   return .immediately
        case "interactively": return .interactively
        case "never":         return .never
        default:              return .automatic
        }
    }
}

File: src/modifiers/scrollDismissesKeyboard.ts
import { createModifier } from './createModifier';

/// Maps to SwiftUI's `ScrollDismissesKeyboardMode`.
export type ScrollDismissesKeyboardMode =
  | 'automatic'
  | 'immediately'
  | 'interactively'
  | 'never';

export interface ScrollDismissesKeyboardParams {
  mode: ScrollDismissesKeyboardMode;
}

/// SwiftUI `.scrollDismissesKeyboard(_:)`. Configures how scrolling
/// dismisses the keyboard. Gated to watchOS 10+ natively; a no-op on
/// older systems.
export function scrollDismissesKeyboard(
  mode: ScrollDismissesKeyboardMode
): ReturnType<typeof createModifier>;
export function scrollDismissesKeyboard(
  params: ScrollDismissesKeyboardParams
): ReturnType<typeof createModifier>;
export function scrollDismissesKeyboard(
  a: ScrollDismissesKeyboardMode | ScrollDismissesKeyboardParams
) {
  if (typeof a === 'string') {
    return createModifier('scrollDismissesKeyboard', { mode: a });
  }
  return createModifier('scrollDismissesKeyboard', a);
}
